From 4585e41b8ae2d8b972bd4655a67f20f441011ba7 Mon Sep 17 00:00:00 2001
From: Takayuki 'January June' Suwa <jjsuwa_sys3175@yahoo.co.jp>
Date: Sun, 2 Jul 2023 02:20:08 +0900
Subject: [PATCH] xtensa: The use of CLAMPS instruction also requires
 TARGET_MINMAX, as well as TARGET_CLAMPS

Because both smin and smax requiring TARGET_MINMAX are essential to the
RTL representation.

gcc/ChangeLog:

	* config/xtensa/xtensa.c (xtensa_match_CLAMPS_imms_p):
	Simplify.
	* config/xtensa/xtensa.md (*xtensa_clamps):
	Add TARGET_MINMAX to the condition.
---
 gcc/config/xtensa/xtensa.c  | 7 ++-----
 gcc/config/xtensa/xtensa.md | 4 ++--
 2 files changed, 4 insertions(+), 7 deletions(-)

diff --git a/gcc/config/xtensa/xtensa.c b/gcc/config/xtensa/xtensa.c
index ac9f79d74..31b3f376e 100644
--- a/gcc/config/xtensa/xtensa.c
+++ b/gcc/config/xtensa/xtensa.c
@@ -2640,11 +2640,8 @@ xtensa_emit_add_imm (rtx dst, rtx src, HOST_WIDE_INT imm, rtx scratch,
 bool
 xtensa_match_CLAMPS_imms_p (rtx cst_max, rtx cst_min)
 {
-  int max, min;
-
-  return IN_RANGE (max = exact_log2 (-INTVAL (cst_max)), 7, 22)
-	 && IN_RANGE (min = exact_log2 (INTVAL (cst_min) + 1), 7, 22)
-	 && max == min;
+  return IN_RANGE (exact_log2 (-INTVAL (cst_max)), 7, 22)
+	 && (INTVAL (cst_max) + INTVAL (cst_min)) == -1;
 }
 
 
diff --git a/gcc/config/xtensa/xtensa.md b/gcc/config/xtensa/xtensa.md
index 035d6214d..feffb87be 100644
--- a/gcc/config/xtensa/xtensa.md
+++ b/gcc/config/xtensa/xtensa.md
@@ -522,7 +522,7 @@
 	(smax:SI (smin:SI (match_operand:SI 1 "register_operand" "r")
 			  (match_operand:SI 2 "const_int_operand" "i"))
 		 (match_operand:SI 3 "const_int_operand" "i")))]
-  "TARGET_CLAMPS
+  "TARGET_MINMAX && TARGET_CLAMPS
    && xtensa_match_CLAMPS_imms_p (operands[3], operands[2])"
   "#"
   "&& 1"
@@ -540,7 +540,7 @@
 	(smin:SI (smax:SI (match_operand:SI 1 "register_operand" "r")
 			  (match_operand:SI 2 "const_int_operand" "i"))
 		 (match_operand:SI 3 "const_int_operand" "i")))]
-  "TARGET_CLAMPS
+  "TARGET_MINMAX && TARGET_CLAMPS
    && xtensa_match_CLAMPS_imms_p (operands[2], operands[3])"
 {
   static char result[64];
-- 
2.30.2

